File: app/src/main/java/me/manabu/utils/NavigationDrawer.java
package me.manabu.utils;


import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.mikepenz.fontawesome_typeface_library.FontAwesome;
import com.mikepenz.materialdrawer.AccountHeader;
import com.mikepenz.materialdrawer.AccountHeaderBuilder;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;
import com.mikepenz.materialdrawer.model.DividerDrawerItem;
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem;
import com.mikepenz.materialdrawer.model.ProfileDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem;
import com.mikepenz.materialdrawer.util.AbstractDrawerImageLoader;
import com.mikepenz.materialdrawer.util.DrawerImageLoader;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import jp.wasabeef.fresco.processors.internal.FastBlur;
import me.manabu.R;
import me.manabu.activities.DecksActivity;
import me.manabu.activities.MainActivity;
import me.manabu.activities.SettingsActivity;

public class NavigationDrawer {

    private Activity activity;
    private Toolbar toolbar;

    private Drawer drawer;

    private AccountHeader accountHeader;
    private Drawable headerDrawable;
    private int BLUR = 5;

    private PrimaryDrawerItem main_page;
    private PrimaryDrawerItem decks;
    private PrimaryDrawerItem settings;
    private PrimaryDrawerItem logout;
    private ProfileDrawerItem profileDrawerItem;
    private static final DividerDrawerItem divider = new DividerDrawerItem();

    public NavigationDrawer(Activity activity, Toolbar toolbar) {
        this.activity = activity;
        this.toolbar = toolbar;

        initImageLoader();
        initItems();
        loadUserPictures(activity);
        initAndBuildDrawer();
        setNowItem();
    }

    private void initAndBuildDrawer() {
        drawer = new DrawerBuilder()
                .withActivity(activity)
                .withToolbar(toolbar)
                .withAccountHeader(accountHeader)
                .addDrawerItems(
                        main_page,
                        decks
                )
                .withCloseOnClick(true)
                .withHeaderPadding(true)
                .withStickyFooterShadow(false)
                .withStickyFooterDivider(true)
                .addStickyDrawerItems(
                        settings,
                        logout
                ).build();
    }

    private void initImageLoader() {
        DrawerImageLoader.init(new AbstractDrawerImageLoader() {
            @Override
            public void set(ImageView imageView, Uri uri, Drawable placeholder, String tag) {
                Picasso.with(imageView.getContext()).load(uri).placeholder(placeholder).into(imageView);
            }

            @Override
            public void cancel(ImageView imageView) {
                Picasso.with(imageView.getContext()).cancelRequest(imageView);
            }
        });
    }

    private void initItems() {
        headerDrawable = activity.getDrawable(R.drawable.header);

        main_page = new PrimaryDrawerItem()
                .withIcon(FontAwesome.Icon.faw_home)
                .withName(R.string.drawer_item_home)
                .withOnDrawerItemClickListener(getClassClicker(MainActivity.class));

        decks = new PrimaryDrawerItem()
                .withIcon(FontAwesome.Icon.faw_list_alt)
                .withName(R.string.drawer_item_decks)
                .withOnDrawerItemClickListener(getClassClicker(DecksActivity.class));

        settings = new PrimaryDrawerItem()
                .withIcon(FontAwesome.Icon.faw_cog)
                .withName(R.string.drawer_item_settings);
                //.withOnDrawerItemClickListener(getClassClicker(SettingsActivity.class));

        logout = new PrimaryDrawerItem()
                .withIcon(FontAwesome.Icon.faw_sign_out_alt)
                .withName(R.string.drawer_item_logout)
                .withOnDrawerItemClickListener((view, position, drawerItem) -> {
                    AuthUtils.signOut(activity);
                    return true;
                });

        profileDrawerItem = new ProfileDrawerItem()
                .withName(AuthUtils.getAccount().getDisplayName())
                .withEmail(AuthUtils.getAccount().getEmail())
                .withIcon(AuthUtils.getAccount().getPhotoUrl());

        accountHeader = new AccountHeaderBuilder()
                .withActivity(activity)
                .withHeaderBackground(headerDrawable)
                .withProfileImagesClickable(false)
                .withSelectionListEnabled(false)
                .addProfiles(profileDrawerItem)
                .build();
    }

    private void loadUserPictures(Activity activity) {
        Picasso.with(activity).load(AuthUtils.getAccount().getPhotoUrl()).into(new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                //Once we get image - update header and variable
                headerDrawable = new BitmapDrawable(activity.getResources(), FastBlur.blur(bitmap, BLUR, false));
                accountHeader.setBackground(headerDrawable);
            }

            @Override
            public void onBitmapFailed(Drawable errorDrawable) {
                Log.d("Navigation Drawer", "Background loading failed!");
            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {
            }
        });
    }

    private void setNowItem() {
        if (activity.getClass() == MainActivity.class) {
            setCurrentItem(main_page);
        }
        if (activity.getClass() == DecksActivity.class) {
            setCurrentItem(decks);
        }
    }

    private void setCurrentItem(PrimaryDrawerItem iDrawerItem) {
        iDrawerItem.withSetSelected(false)
                .withOnDrawerItemClickListener(null);
        drawer.setSelection(iDrawerItem, false);
    }

    private Drawer.OnDrawerItemClickListener getClassClicker(Class className) {
        return (view, position, drawerItem) -> {
            Intent intent = new Intent(activity, className);
            activity.startActivity(intent);
            return true;

        };
    }

    public Drawer getDrawer() {
        return drawer;
    }

    public Toolbar getToolbar() {
        return toolbar;
    }
}

File: app/src/main/java/me/manabu/modules/NavigationDrawer.kt
package me.manabu.modules

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.support.v4.app.Fragment
import co.zsmb.materialdrawerkt.builders.*
import co.zsmb.materialdrawerkt.draweritems.badgeable.primaryItem
import co.zsmb.materialdrawerkt.draweritems.profile.profile
import co.zsmb.materialdrawerkt.imageloader.drawerImageLoader
import com.mikepenz.fontawesome_typeface_library.FontAwesome
import com.mikepenz.materialdrawer.AccountHeader
import com.mikepenz.materialdrawer.Drawer
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.ProfileDrawerItem
import com.mikepenz.materialdrawer.util.DrawerUIUtils
import com.squareup.picasso.Picasso
import jp.wasabeef.fresco.processors.internal.FastBlur
import me.manabu.R
import me.manabu.activities.MainActivity
import me.manabu.activities.fragments.DecksFragment
import me.manabu.activities.fragments.MainFragment
import me.manabu.activities.fragments.SettingsFragment
import org.jetbrains.anko.toast

object NavigationDrawer {

    private const val headerBlur = 5

    private lateinit var accountHeader: AccountHeader
    private var headerDrawable: Drawable? = null

    private lateinit var mainPage: PrimaryDrawerItem
    lateinit var decksPage: PrimaryDrawerItem
    private lateinit var settings: PrimaryDrawerItem
    private lateinit var logout: PrimaryDrawerItem
    private lateinit var userProfile: ProfileDrawerItem


    fun build(context: MainActivity): Drawer {
        if (!CurrentUser.isSignedIn()) {
            throw ExceptionInInitializerError("Account still not initialized \nWhy are you actually trying to build a drawer, when we're still not signed in?")
        }

        initImageLoader()

        return context.drawer {
            toolbar = context.toolbar

            closeOnClick = true
            headerPadding = true
            stickyFooterShadow = false
            stickyFooterDivider = true

            accountHeader = accountHeader {
                profileImagesClickable = false
                selectionListEnabled = false
                backgroundDrawable = getBg(context)

                userProfile = profile(CurrentUser.getAccount().displayName!!, CurrentUser.getAccount().email!!) {
                    iconUrl = CurrentUser.getAccount().photoUrl.toString()
                }
            }


            //Elements
            mainPage = primaryItem(R.string.drawer_item_home) {
                iicon = FontAwesome.Icon.faw_home
                onClick { _ ->
                    context.changeFragment(MainFragment())
                    false
                }
            }
            decksPage = primaryItem(R.string.drawer_item_decks) {
                iicon = FontAwesome.Icon.faw_list_alt
                onClick { _ ->
                    context.changeFragment(DecksFragment())
                    false
                }
            }
            footer {
                settings = primaryItem(R.string.drawer_item_settings) {
                    iicon = FontAwesome.Icon.faw_cog
                    onClick { _ ->
                        context.changeFragment(SettingsFragment())
                        false
                    }
                }
                logout = primaryItem(R.string.drawer_item_logout) {
                    iicon = FontAwesome.Icon.faw_sign_out_alt
                    onClick { _ ->
                        CurrentUser.signOut(context)
                        false
                    }
                }

            }
        }
    }

    private fun getBg(activity: Activity): Drawable {
        if (headerDrawable != null)
            return headerDrawable as Drawable

        return activity.getDrawable(R.drawable.header)
    }

    private fun initImageLoader() {
        drawerImageLoader {
            placeholder { ctx, _ ->
                DrawerUIUtils.getPlaceHolder(ctx)
            }
            set { imageView, uri, placeholder, _ ->
                Picasso.with(imageView.context)
                        .load(uri)
                        .placeholder(placeholder)
                        .into(imageView)
            }
            cancel { imageView ->
                Picasso.with(imageView.context)
                        .cancelRequest(imageView)
            }
        }
    }

    fun loadUserBackground(activity: Activity) {
        Picasso.with(activity).load(CurrentUser.getAccount().photoUrl).into(object : com.squareup.picasso.Target {
            override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                //Once we getAccount image - update header and variable
                headerDrawable = BitmapDrawable(activity.resources, FastBlur.blur(bitmap, headerBlur, false))
                accountHeader.setBackground(headerDrawable)
            }

            override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}
            override fun onBitmapFailed(errorDrawable: Drawable?) {
                activity.toast("Navigation Drawer: Can't load user background")
            }
        })
    }
}